src/proc/occlusion-filter.cpp

```diff
@@ -45,51 +81,116 @@ namespace librealsense
+        {
+            depth_planes[0] = (byte*)alloc.allocate(depth.get_bytes_per_pixel() * frame_size);
+            int bpp = depth.get_bytes_per_pixel();
+            rotate_image_counterclockwise<2>(depth_planes, (const byte*)(depth.get_data()), points_width, points_height);
+
+            // scan depth frame after clockwise rotation: check if there is a significant jump between adjacen pixels in Z-axis (depth), it means there could be occlusion.
+            // scan over a small window so that points depth is the same
+            // save suspected points and run occlusion-invalidation vertical scan only on them
+            // after rotation : height = points_width , width = points_height
+
+            auto rotated_depth_width = _depth_intrinsics->height;
+            auto rotated_depth_height = _depth_intrinsics->width;
+
+            for (int i = 0; i < rotated_depth_height; i++)
+            {
+                for (int j = 0; j < rotated_depth_width - VERTICAL_SCAN_WINDOW_SIZE; j = j + VERTICAL_SCAN_WINDOW_SIZE)
+                {
+                    // before rotation: occlusion detected in the positive direction of Y
+                    // after counterclockwise rotation : scan from right to left (positive direction of X) to detect occlusion 
+                    // compare depth each pixel only with the pixel on its right (i,j+1) 
+                    auto index = i * rotated_depth_width + j;
+                    auto uv_i = j;
+                    auto uv_j = rotated_depth_height - i;
+                    auto uv_index = uv_i * rotated_depth_height + uv_j; // 90 degrees rotation transform from rotated depth
+                    auto index_right = index + 1;
+                    float diff_right = abs((depth_planes[0])[index] - (depth_planes[0])[index_right]);
+
+                    if ((diff_right > DEPTH_OCCLUSION_THRESHOLD))
                     {
-                        maxInLine = pixels_ptr->x;
-                        maxZ = points->z;
-                        if (occDilationLeft > 0)
+                        pixels_ptr = pix_coord.data() + uv_index;
+                        points_ptr = points + uv_index;
+                        uv_map_ptr = uv_map + uv_index;
+                        maxInLine = -1;
+                        maxZ = 0;
+                        int occDilationUp = 0;
+
+                        for (size_t y = 0; y <= VERTICAL_SCAN_WINDOW_SIZE; ++y)
                         {
-                            uv_map->x = 0.f;
-                            uv_map->y = 0.f;
-                            occDilationLeft--;
+                            if (((pixels_ptr + y * points_width)->y < maxInLine))
+                            {
+                                *(points_ptr + y * points_width) = { 0.f, 0.f, 0.f };
+                            }
+                            else
+                            {
+                                maxInLine = (pixels_ptr + y * points_width)->y;
```

Comment:
When reaching new max, break out of the loop early

Reply:
I checked with breaking the loop, many occluded points are missed because pixel depth is not always a good indication for occluded points, sometimes close pixels have same depth but both are occluded, so it is better to keep it like this.